src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/DecompressionHandler.cs

```diff
@@ -224,12 +223,203 @@ public DeflateDecompressedContent(HttpContent originalContent)
+                private Stream? _decompressionStream;
+
+                public ZLibOrDeflateStream(Stream stream) => _stream = new PeekFirstByteReadStream(stream);
+
+                protected override void Dispose(bool disposing)
+                {
+                    if (disposing)
+                    {
+                        _decompressionStream?.Dispose();
+                        _stream.Dispose();
+                    }
+                    base.Dispose(disposing);
+                }
+
+                public override bool CanRead => true;
+                public override bool CanWrite => false;
+                public override ValueTask WriteAsync(ReadOnlyMemory<byte> buffer, CancellationToken cancellationToken) => throw new NotSupportedException();
+
+                // On the first read request, peek at the first nibble of the response. If it's an 8, use ZLibStream, otherwise
+                // use DeflateStream. This heuristic works because we're deciding only between raw deflate and zlib wrapped around
+                // deflate, in which case the first nibble will always be 8 for zlib and never be 8 for deflate.
+                // https://stackoverflow.com/a/37528114 provides an explanation for why.
+
+                public override int Read(Span<byte> buffer)
+                {
+                    if (_decompressionStream is null)
+                    {
+                        int firstByte = _stream.PeekFirstByte();
+                        _decompressionStream = CreateDecompressionStream(firstByte, _stream);
+                    }
+
+                    return _decompressionStream.Read(buffer);
+                }
+
+                public override ValueTask<int> ReadAsync(Memory<byte> buffer, CancellationToken cancellationToken)
+                {
+                    if (_decompressionStream is null)
+                    {
+                        return CreateAndReadAsync(this, buffer, cancellationToken);
+
+                        static async ValueTask<int> CreateAndReadAsync(ZLibOrDeflateStream thisRef, Memory<byte> buffer, CancellationToken cancellationToken)
+                        {
+                            int firstByte = await thisRef._stream.PeekFirstByteAsync(cancellationToken).ConfigureAwait(false);
+                            thisRef._decompressionStream = CreateDecompressionStream(firstByte, thisRef._stream);
+                            return await thisRef._decompressionStream.ReadAsync(buffer, cancellationToken).ConfigureAwait(false);
+                        }
+                    }
+
+                    return _decompressionStream.ReadAsync(buffer, cancellationToken);
+                }
+
+                public override Task CopyToAsync(Stream destination, int bufferSize, CancellationToken cancellationToken)
+                {
+                    ValidateCopyToArguments(destination, bufferSize);
+                    return Core(destination, bufferSize, cancellationToken);
```

Comment:
So that any exceptions from the argument validation performed by ValidateCopyToArguments are propagated synchronously rather than getting stored into the returned task.